testing/src/test/java/io/stargate/it/cql/BatchStatementTest.java

```diff
@@ -0,0 +1,346 @@
+        SimpleStatement.builder(
+                String.format(
+                    "INSERT INTO test (k0, k1, v) values ('%s', ? , ?)", name.getDisplayName()))
+            .build();
+    PreparedStatement preparedStatement = session.prepare(insert);
+
+    for (int i = 0; i < batchCount; i++) {
+      if (i % 2 == 1) {
+        SimpleStatement simpleInsert =
+            SimpleStatement.builder(
+                    String.format(
+                        "INSERT INTO test (k0, k1, v) values ('%s', ?, ?)", name.getDisplayName()))
+                .addPositionalValues(i, i + 1)
+                .build();
+        builder.addStatement(simpleInsert);
+      } else {
+        builder.addStatement(preparedStatement.bind(i, i + 1));
+      }
+    }
+
+    BatchStatement batchStatement = builder.build();
+    session.execute(batchStatement);
+
+    verifyBatchInsert(name);
+  }
+
+  @Test
+  public void should_execute_cas_batch(TestInfo name) {
+    // Build a batch with CAS operations on the same partition.
+    BatchStatementBuilder builder = BatchStatement.builder(DefaultBatchType.UNLOGGED);
+    SimpleStatement insert =
+        SimpleStatement.builder(
+                String.format(
+                    "INSERT INTO test (k0, k1, v) values ('%s', ? , ?) IF NOT EXISTS",
+                    name.getDisplayName()))
+            .build();
+    PreparedStatement preparedStatement = session.prepare(insert);
+
+    for (int i = 0; i < batchCount; i++) {
+      builder.addStatement(preparedStatement.bind(i, i + 1));
+    }
+
+    BatchStatement batchStatement = builder.build();
+    ResultSet result = session.execute(batchStatement);
+    assertThat(result.wasApplied()).isTrue();
+
+    verifyBatchInsert(name);
+
+    // re execute same batch and ensure wasn't applied.
+    result = session.execute(batchStatement);
+    assertThat(result.wasApplied()).isFalse();
+  }
+
+  @Test
+  public void should_execute_counter_batch(TestInfo name) {
+    // should be able to do counter increments in a counter batch.
+    BatchStatementBuilder builder = BatchStatement.builder(DefaultBatchType.COUNTER);
+
+    for (int i = 1; i <= 3; i++) {
+      SimpleStatement insert =
+          SimpleStatement.builder(
+                  String.format(
+                      "UPDATE counter%d set c = c + %d where k0 = '%s'",
+                      i, i, name.getDisplayName()))
+              .build();
+      builder.addStatement(insert);
+    }
+
+    BatchStatement batchStatement = builder.build();
+    session.execute(batchStatement);
+
+    for (int i = 1; i <= 3; i++) {
+      ResultSet result =
+          session.execute(
+              String.format("SELECT c from counter%d where k0 = '%s'", i, name.getDisplayName()));
+
+      List<Row> rows = result.all();
+      assertThat(rows).hasSize(1);
+
+      Row row = rows.iterator().next();
+      assertThat(row.getLong("c")).isEqualTo(i);
+    }
+  }
+
+  private void doCounterIncrement(TestInfo name) {
```

Comment:
I was concerned that moving all of this inline would obscure what was being tested... so I decided for a well-chosen private method name instead.